Python/instrumentation.c

```diff
@@ -1514,25 +1579,28 @@ allocate_instrumentation_data(PyCodeObject *code)
 {
 
     if (code->_co_monitoring == NULL) {
-        code->_co_monitoring = PyMem_Malloc(sizeof(_PyCoMonitoringData));
-        if (code->_co_monitoring == NULL) {
+        _PyCoMonitoringData *monitoring = PyMem_Malloc(sizeof(_PyCoMonitoringData));
+        if (monitoring == NULL) {
             PyErr_NoMemory();
             return -1;
         }
-        code->_co_monitoring->local_monitors = (_Py_LocalMonitors){ 0 };
-        code->_co_monitoring->active_monitors = (_Py_LocalMonitors){ 0 };
-        code->_co_monitoring->tools = NULL;
-        code->_co_monitoring->lines = NULL;
-        code->_co_monitoring->line_tools = NULL;
-        code->_co_monitoring->per_instruction_opcodes = NULL;
-        code->_co_monitoring->per_instruction_tools = NULL;
+        monitoring->local_monitors = (_Py_LocalMonitors){ 0 };
+        monitoring->active_monitors = (_Py_LocalMonitors){ 0 };
+        monitoring->tools = NULL;
+        monitoring->lines = NULL;
+        monitoring->line_tools = NULL;
+        monitoring->per_instruction_opcodes = NULL;
+        monitoring->per_instruction_tools = NULL;
+        FT_ATOMIC_STORE_PTR_RELEASE(code->_co_monitoring, monitoring);
```

Comment:
We actually do lock the code object on initialization, but the current implementation was trying to avoid locking the code object on reads.  The readers of `_co_monitoring` are all using acquire semantics so if there is a value there then they'll also pick all of the writes to the data which it is pointing to as well.  The alternative here is to either lock the code object on reads as well or to stop the world on `_Py_Instrument` calls which seems to have disproportionately compared to calls to actually enable monitoring.